source/server/admin/prometheus_stats.cc

```diff
@@ -146,6 +148,14 @@ std::string generateNumericOutput(const StatType& metric,
   return fmt::format("{0}{{{1}}} {2}\n", prefixed_tag_extracted_name, tags, metric.value());
 }
 
+std::string generateTextReadoutOutput(const Stats::TextReadout& text_readout,
+                                      const std::string& prefixed_tag_extracted_name) {
+  auto tags = std::vector(text_readout.tags());
+  tags.push_back(Stats::Tag{"text_value", text_readout.value()});
+  const std::string formattedTags = PrometheusStatsFormatter::formattedTags(tags);
+  return fmt::format("{0}{{{1}}} 1\n", prefixed_tag_extracted_name, formattedTags);
```

Comment:
do we need to worry about quoting text_readout.value() and escaping it, in case it's got some delimiters, newlines, quotes in it, itself?

A reference to Prometheus doc about how to inject arbitrary text would be good here.

Reply:
The [documentation](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) says that a label value may contain any Unicode character.

> A reference to Prometheus doc about how to inject arbitrary text would be good here.
Do you suggest to add a comment in the code?

Reply:
Yes definitely add comments referencing the Prometheus doc link, and add examples for this use-case.

And I can see from PrometheusStatsFormatter::formattedTags that the values are enclosed in double-quotes. What's your plan if a value includes double-quotes?

Reply:
Ok, I will add a comment.

> What's your plan if a value includes double-quotes?

To be honest, I didn't consider this case and as far as this case is valid for prometheus I don't see a need to handle it. Do you disagree with me?

Reply:
It's really not up to me; it's up to Prometheus, but yes! :)

If you have a text-readout whose .value() contains a single double-quote, IIUC you'll get Prometheus formatting which contains an odd number of double-quotes. I have never used Prometheus but I'd bet it would fail to parse, or do something worse.

Reply:
Ok, thank you very much for this advice. I see two options:
1. Remove fixed set of characters (single quotes, double quotes, new lines and so on) from text-readout value if exist.
2. Match text-readout value with regex with allowed characters and skip if does not match. Something like this:
```
if (!std::match_regex(textReadout.value(), std::regex("some regex")) {
  return "";
}
```
What do you think?

Reply:
that would avoid corrupting the Prometheus output, and may be good enough.

Does Prometheus define a way to escape quotes or other special characters, e.g. `\042` so you can just sanitize without potentially losing data?

Reply:
> Does Prometheus define a way to escape quotes or other special characters, e.g. \042 so you can just sanitize without potentially losing data?

I don't know, I will dive deeper into this topic and I let you know.

Reply:
It seems that I don't have to care about escaping, because this is already in progress in #18978. I will wait for this to be ready.

Reply:
Ok, the PR that I mentioned above was merged into the main branch and it solved the problem of escaping problematic characters, so I think that there is no longer need to change anything here. Do you agree with me?

Reply:
lgtm -- I assume that was previously verified as the correct syntax for escaping in Prometheus. The significance of this PR is that I suspect this enables arbitrary to text to flow into Prometheus tags in a way that wasn't likely before.